src/client/clientmap.cpp

```diff
@@ -299,213 +291,290 @@ void ClientMap::updateDrawList()
 	// if (occlusion_culling_enabled && m_control.show_wireframe)
 	// 	occlusion_culling_enabled = porting::getTimeS() & 1;
 
-	std::queue<v3s16> blocks_to_consider;
+	// Set of mesh holding blocks
+	std::set<v3s16> shortlist;
 
-	v3s16 camera_mesh = mesh_grid.getMeshPos(camera_block);
-	v3s16 camera_cell = mesh_grid.getCellPos(camera_block);
+	/*
+	 When range_all is enabled, enumerate all blocks visible in the
+	 frustum and display them.
+	 */
+	if (m_control.range_all) {
+		MapBlockVect sectorblocks;
 
-	// Bits per block:
-	// [ visited | 0 | 0 | 0 | 0 | Z visible | Y visible | X visible ]
-	MapBlockFlags meshes_seen(mesh_grid.getCellPos(p_blocks_min), mesh_grid.getCellPos(p_blocks_max) + 1);
+		for (auto &sector_it : m_sectors) {
+			MapSector *sector = sector_it.second;
+			if (!sector)
+				continue;
 
-	// Start breadth-first search with the block the camera is in
-	blocks_to_consider.push(camera_mesh);
-	meshes_seen.getChunk(camera_cell).getBits(camera_cell) = 0x07; // mark all sides as visible
+			sectorblocks.clear();
+			sector->getBlocks(sectorblocks);
 
-	std::set<v3s16> shortlist;
+			// Loop through blocks in sector
+			for (MapBlock *block : sectorblocks) {
+				MapBlockMesh *mesh = block->mesh;
 
-	// Recursively walk the space and pick mapblocks for drawing
-	while (blocks_to_consider.size() > 0) {
+				// Calculate the coordinates for range and frustum culling
+				v3f mesh_sphere_center;
+				f32 mesh_sphere_radius;
 
-		v3s16 block_coord = blocks_to_consider.front();
-		blocks_to_consider.pop();
+				v3s16 block_pos_nodes = block->getPos() * MAP_BLOCKSIZE;
 
-		v3s16 cell_coord = mesh_grid.getCellPos(block_coord);
-		auto &flags = meshes_seen.getChunk(cell_coord).getBits(cell_coord);
+				if (mesh) {
+					mesh_sphere_center = intToFloat(block_pos_nodes, BS)
+							+ mesh->getBoundingSphereCenter();
+					mesh_sphere_radius = mesh->getBoundingRadius();
+				} else {
+					mesh_sphere_center = intToFloat(block_pos_nodes, BS)
+							+ v3f((MAP_BLOCKSIZE * 0.5f - 0.5f) * BS);
+					mesh_sphere_radius = 0.0f;
+				}
```

Comment:
Not needed. As blocks / meshes are not traversed recursively, empty blocks can just be skipped.

This also means that each mesh will only be added once, so `shortlist` could be a simple vector.
(It looks to me like de-duplicating isn't needed in finite view range either, but that's something for later refactor I guess.)

Reply:
You still want to add them to `m_keeplist`, so can't skip